GitUI/CommitInfo/RefsFormatter.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using GitCommands;
+using ResourceManager;
+
+namespace GitUI.CommitInfo
+{
+    public sealed class RefsFormatter : ITranslate
+    {
+        /// <summary>
+        /// The number of displayed refs if the list is limited.
+        ///
+        /// If limited, the line "[Show all]" and an empty line are added.
+        /// Hence the list needs to be limited only if it exceeds MaximumDisplayedRefsIfLimited + 2.
+        /// </summary>
+        private const int MaximumDisplayedRefsIfLimited = 10;
+
+        private static readonly TranslationString _containedInBranches = new TranslationString("Contained in branches:");
+        private static readonly TranslationString _containedInNoBranch = new TranslationString("Contained in no branch");
+        private static readonly TranslationString _containedInTags = new TranslationString("Contained in tags:");
+        private static readonly TranslationString _containedInNoTag = new TranslationString("Contained in no tag");
+        private static readonly ILinkFactory _linkFactory = new LinkFactory();
+
+        public static string FormatBranches(IEnumerable<string> branches, bool showAsLinks, bool limit)
+        {
+            var links = new List<string>();
+            bool truncated = false;
+
+            const string remotesPrefix = "remotes/";
+
+            // Include local branches if explicitly requested or when needed to decide whether to show remotes
+            bool getLocal = AppSettings.CommitInfoShowContainedInBranchesLocal ||
+                            AppSettings.CommitInfoShowContainedInBranchesRemoteIfNoLocal;
+
+            // Include remote branches if requested
+            bool getRemote = AppSettings.CommitInfoShowContainedInBranchesRemote ||
+                             AppSettings.CommitInfoShowContainedInBranchesRemoteIfNoLocal;
+            bool allowLocal = AppSettings.CommitInfoShowContainedInBranchesLocal;
+            bool allowRemote = getRemote;
+
+            foreach (var branch in branches)
+            {
+                string noPrefixBranch = branch;
+                bool branchIsLocal;
+                if (getLocal && getRemote)
+                {
+                    // "git branch -a" prefixes remote branches with "remotes/"
+                    // It is possible to create a local branch named "remotes/origin/something"
+                    // so this check is not 100% reliable.
+                    // This shouldn't be a big problem if we're only displaying information.
+                    branchIsLocal = !branch.StartsWith(remotesPrefix);
```

Comment:
improve on comment
// This could be solved by listing local and remote branches separetly